### DISCUSSION TITLE
Rye performance

### DISCUSSION THREAD

**Discussion by ipmb** [astral-sh/rye#84] [General]
Hey @mitsuhiko! I'm excited about Rye. I've been a `pip-tools` user for a while. I like that rye is trying to fit within existing Python standards and not straying too far where they don't exist yet. Generating requirements files that can be installed without any special tooling beyond `pip` is great.

I maintain https://lincolnloop.github.io/python-package-manager-shootout/ and plugged in Rye to see how it stacked up. I realize its early days and this may not be a priority yet, but wanted to call out a few things I noticed from [the results here](https://github.com/lincolnloop/python-package-manager-shootout/actions/runs/4853217930#summary-13150224765)

1. Install is significantly slower. Expected given it's currently built from source. I assume that once things stabilize, binaries will be available.
2. Based on the speed, adding a package seems like it only updates `pyproject.toml`. I believe the other tools update the lock file as well in this step. Was this intentional?
3. The first time it runs, I see this in the logs which doesn't really make it a "fair fight" vs other tooling which does this during the "install" phase. Is there a way to trigger this before the environment is initialized?

    ```
    Bootstrapping rye internals
    Upgrading pip
    ```

4. It looks like running `rye sync` regenerates the lock file even if it already exists? 
![image](https://user-images.githubusercontent.com/319156/235494888-79566e2e-96d3-4dde-b0e4-57125cddd1f6.png)
5. `lock` generates a dev lock file even if no dev dependencies exist

**Comment by cnpryer** [comment]
FWIW performance is an area I'd love to contribute to here, but yea this is super super early.

A lot of these results should be expected. At this stage the idea is to just evaluate `rye`'s fit in the Python ecosystem.

As an outsider to this project, take my thoughts with a grain of salt.

When optimizing performance is fully on the table there are really these areas to focus on:
- requirements parsing
- source caching
- dependency resolving
- packaging/installing/unpacking operations
